File: src/queue/linkedListQueue.c
/*Queue - Linked List implementation*/
#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
  int data;
  struct Node *next;
} Node;

// to enqueue an integer
void enqueue(Node **head, int data)
{
  // Creating a memory for newNode
  Node *newNode = malloc(sizeof(Node));
  if (!newNode)
    return;

  newNode->data = data;
  newNode->next = *head;

  *head = newNode;
}
// to dequeue an integer
int dequeue(Node **head)
{
  Node *current, *prev = NULL;
  int retval = -1;

  if (*head == NULL)
    return -1;

  current = *head;
  while (current->next != NULL)
  {
    prev = current;
    current = current->next;
  }

  retval = current->data;
  free(current);

  if (prev)
    prev->next = NULL;
  else
    *head = NULL;

  return retval;
}

void print_list(Node *head)
{
  Node *current = head;

  while (current != NULL)
  {
    printf("%d\n", current->data);
    current = current->next;
  }
}

int main()
{
  Node *head = NULL;
  int ret;

  enqueue(&head, 11);
  enqueue(&head, 22);
  enqueue(&head, 33);
  enqueue(&head, 44);

  print_list(head);

  while ((ret = dequeue(&head)) > 0)
  {
    printf("dequeued %d\n", ret);
  }
  printf("done. head=%p\n", head);

  return 0;
}
/*
src/queue> gcc ./linkedListQueue.c
src/queue> a.out

4
33
22
11
dequeued 11
dequeued 22
dequeued 33
dequeued 44
done. head=0x0
*/

File: src/queue/linkedListQueue.ts
/* JavaScript program for linked-list implementation of queue */

// Node this interface
interface INode<T> {
  data: T
  next: INode<T> | null
}
interface ILinkedListQueue<T> {
  // size: () => number
  front: () => INode<T> | null
  rear: () => INode<T> | null
  enqueue: (x: T) => boolean
  dequeue: () => boolean
  // isEmpty: () => boolean
}
function LinkedListQueue<T>(this: ILinkedListQueue<T>) {
  let front: INode<T> | null = null
  let rear: INode<T> | null = null
  this.front = function () {
    return front
  }
  this.rear = function () {
    return rear
  }
  const Node = function (this: INode<T>, data: T) {
    this.data = data
    this.next = null
  }
  this.enqueue = function (element: T) {
    // Create a new LL node
    let temp = new Node(element)
    // If queue is empty, then new node is front and rear both
    if (rear == null) {
      front = rear = temp
      return true
    }
    // Add the new node at the end of queue and change rear
    rear.next = temp
    rear = temp
    return true
  }
  this.dequeue = function () {
    // If queue is empty, return NULL.
    if (front == null) return false // Store previous front and move front one node ahead
    let temp = front
    front = front.next // If front becomes NULL, then change rear also as NULL
    if (front == null) rear = null
    return true
  }
}

// Export functions
export { LinkedListQueue }

// Code to test the implementation.
// const myLLQueue = new LinkedListQueue<number>()
// console.log(myLLQueue.enqueue(10))
// console.log(myLLQueue.enqueue(20))
// console.log(myLLQueue.enqueue(30))
// console.log(myLLQueue.dequeue())
// console.log(myLLQueue.front())
